File: lib/pages/weather/airkorea_api.dart
// lib/pages/air/airkorea_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../../env.dart';

/// 실시간 시/도별 측정소 대기질 데이터
class AirKoreaApi {
  static const _host = 'apis.data.go.kr';
  // 시도별 실시간 측정 정보
  static const _path = '/B552584/ArpltnInforInqireSvc/getCtprvnRltmMesureDnsty';

  static Future<List<AirQuality>> fetchStationData({
    String sidoName = '서울',
    int numOfRows = 100,
    int pageNo = 1,
  }) async {
    final serviceKey = Env.AIRKOREA_SERVICE_KEY;
    if (serviceKey.isEmpty) {
      throw Exception('api key 발급 필요');
    }

    final query = {
      'serviceKey': serviceKey,
      'returnType': 'json',
      'sidoName': sidoName,
      'numOfRows': '$numOfRows',
      'pageNo': '$pageNo',
      'ver': '1.3',
    };

    final uri = Uri.https(_host, _path, query);
    final res = await http.get(uri);
    if (res.statusCode != 200) {
      throw Exception('AirKorea HTTP ${res.statusCode}: ${res.body}');
    }

    final Map<String, dynamic> j = json.decode(res.body);
    final body = j['response']?['body'];
    if (body == null) return [];

    final List items = body['items'] ?? [];
    return items
        .map((e) => AirQuality.fromJson(e as Map<String, dynamic>))
        .toList();
  }
}

class AirQuality {
  final String dataTime;

  final String? o3Value;
  final String? no2Value;
  final String? pm10Value;
  final String? pm25Value;

  final String? o3Grade;
  final String? no2Grade;
  final String? pm10Grade;
  final String? pm25Grade;

  final String? pm10Grade1h;
  final String? pm25Grade1h;

  final String? o3Flag;
  final String? no2Flag;
  final String? pm10Flag;
  final String? pm25Flag;

  // 선택: UI 표시에 유용
  final String? stationName;
  final String? sidoName;

  AirQuality({
    required this.dataTime,
    this.o3Value,
    this.no2Value,
    this.pm10Value,
    this.pm25Value,
    this.o3Grade,
    this.no2Grade,
    this.pm10Grade,
    this.pm25Grade,
    this.pm10Grade1h,
    this.pm25Grade1h,
    this.o3Flag,
    this.no2Flag,
    this.pm10Flag,
    this.pm25Flag,
    this.stationName,
    this.sidoName,
  });

  factory AirQuality.fromJson(Map<String, dynamic> j) {
    String? _str(dynamic v) =>
        (v == null || (v is String && v.trim().isEmpty)) ? null : v.toString();

    return AirQuality(
      dataTime: j['dataTime']?.toString() ?? '',
      o3Value: _str(j['o3Value']),
      no2Value: _str(j['no2Value']),
      pm10Value: _str(j['pm10Value']),
      pm25Value: _str(j['pm25Value']),
      o3Grade: _str(j['o3Grade']),
      no2Grade: _str(j['no2Grade']),
      pm10Grade: _str(j['pm10Grade']),
      pm25Grade: _str(j['pm25Grade']),
      pm10Grade1h: _str(j['pm10Grade1h']),
      pm25Grade1h: _str(j['pm25Grade1h']),
      o3Flag: _str(j['o3Flag']),
      no2Flag: _str(j['no2Flag']),
      pm10Flag: _str(j['pm10Flag']),
      pm25Flag: _str(j['pm25Flag']),
      stationName: _str(j['stationName']),
      sidoName: _str(j['sidoName']),
    );
  }
}

File: android/app/src/main/kotlin/com/example/dive_app/api/AirKoreaApi.kt
package com.example.dive_app.api

import android.annotation.SuppressLint
import android.content.Context
import android.location.Geocoder
import android.util.Log
import com.google.android.gms.location.LocationServices
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.util.Locale
import kotlinx.coroutines.tasks.await
import com.example.dive_app.BuildConfig
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

object AirKoreaApi {
    private const val host = "apis.data.go.kr"
    private const val path = "/B552584/ArpltnInforInqireSvc/getCtprvnRltmMesureDnsty"
    private val serviceKey = BuildConfig.AIRKOREA_SERVICE_KEY

    // 🔥 suspend 추가
    suspend fun fetchAirQuality(sidoName: String): JSONObject? {
        return withContext(Dispatchers.IO) {   // ✅ IO 스레드에서 실행
            try {
                val encodedKey = URLEncoder.encode(serviceKey, "UTF-8")
                val encodedSido = URLEncoder.encode(sidoName, "UTF-8")

                val urlStr = "https://apis.data.go.kr$path?" +
                        "serviceKey=$encodedKey&returnType=json&sidoName=$encodedSido" +
                        "&numOfRows=1&pageNo=1&ver=1.3"

                val conn = URL(urlStr).openConnection() as HttpURLConnection
                conn.requestMethod = "GET"

                val res = conn.inputStream.bufferedReader().use { it.readText() }

                Log.d("AirKoreaApi", "raw response: $res")

                JSONObject(res)  // ✅ 이제 결과 JSON 리턴 가능
            } catch (e: Exception) {
                Log.e("AirKoreaApi", "fetchAirQuality 실패", e)
                null
            }
        }
    }

    @SuppressLint("MissingPermission")
    suspend fun fetchAirQualityByLocation(context: Context): JSONObject? {
        return try {
            val fused = LocationServices.getFusedLocationProviderClient(context)
            val location = fused.lastLocation.await()

            val geocoder = Geocoder(context, Locale.KOREA)
            val addr = location?.let {
                geocoder.getFromLocation(it.latitude, it.longitude, 1)
            }
            val adminArea = addr?.firstOrNull()?.adminArea ?: "서울"
            val sidoName = normalizeRegion(adminArea)

            // ✅ 원본 JSON 받아오기
            val rawJson: JSONObject? = fetchAirQuality(sidoName)

            if (rawJson != null) {
                val response = rawJson.getJSONObject("response")
                val body = response.getJSONObject("body")
                val items = body.getJSONArray("items")

                if (items.length() > 0) {
                    val first = items.getJSONObject(0)

                    // ✅ 필요한 값만 추출해서 리턴
                    JSONObject().apply {
                        put("pm10Value", first.optString("pm10Value"))
                        put("pm10Grade", first.optString("pm10Grade1h"))
                        put("pm25Value", first.optString("pm25Value"))
                        put("pm25Grade", first.optString("pm25Grade1h"))
                        put("o3Value", first.optString("o3Value"))
                        put("o3Grade", first.optString("o3Grade"))
                        put("no2Value", first.optString("no2Value"))
                        put("no2Grade", first.optString("no2Grade"))
                    }
                } else {
                    null
                }
            } else {
                null
            }
        } catch (e: Exception) {
            Log.e("AirKoreaApi", "위치 기반 대기질 조회 실패", e)
            null
        }
    }

    private fun normalizeRegion(adminArea: String): String {
        val s = adminArea.replace("특별시", "")
            .replace("광역시", "")
            .replace("특별자치도", "")
            .replace("특별자치시", "")
            .replace("도", "")
            .trim()

        return when {
            s.contains("서울") -> "서울"
            s.contains("부산") -> "부산"
            s.contains("대구") -> "대구"
            s.contains("인천") -> "인천"
            s.contains("광주") -> "광주"
            s.contains("대전") -> "대전"
            s.contains("울산") -> "울산"
            s.contains("세종") -> "세종"
            s.contains("경기") -> "경기"
            s.contains("강원") -> "강원"
            s.contains("충북") -> "충북"
            s.contains("충남") -> "충남"
            s.contains("전북") -> "전북"
            s.contains("전남") -> "전남"
            s.contains("경북") -> "경북"
            s.contains("경남") -> "경남"
            s.contains("제주") -> "제주"
            else -> "서울"
        }
    }
}

